src/Reflection/ReflectionConstant.php

```diff
@@ -0,0 +1,283 @@
+
+namespace Roave\BetterReflection\Reflection;
+
+use Exception;
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Namespace_ as NamespaceNode;
+use Reflector as CoreReflector;
+use Roave\BetterReflection\BetterReflection;
+use Roave\BetterReflection\NodeCompiler\CompileNodeToValue;
+use Roave\BetterReflection\NodeCompiler\CompilerContext;
+use Roave\BetterReflection\Reflection\StringCast\ReflectionConstantStringCast;
+use Roave\BetterReflection\Reflector\Exception\IdentifierNotFound;
+use Roave\BetterReflection\Reflector\Reflector;
+use Roave\BetterReflection\SourceLocator\Located\LocatedSource;
+use Roave\BetterReflection\Util\CalculateReflectionColum;
+use Roave\BetterReflection\Util\GetFirstDocComment;
+use function implode;
+
+class ReflectionConstant implements Reflection, CoreReflector
+{
+    /** @var Reflector */
+    private $reflector;
+
+    /** @var Node\Stmt\Const_|Node\Expr\FuncCall */
+    private $node;
+
+    /** @var LocatedSource */
+    private $locatedSource;
+
+    /** @var NamespaceNode|null */
+    private $declaringNamespace;
+
+    /** @var int|null */
+    private $positionInNode;
+
+    /** @var int|float|mixed[]|string|bool|null const value */
+    private $value;
+
+    /** @var bool */
+    private $valueWasCached = false;
+
+    private function __construct()
+    {
+    }
+
+    /**
+     * Create a ReflectionConstant by name, using default reflectors etc.
+     *
+     * @throws IdentifierNotFound
+     */
+    public static function createFromName(string $constantName) : self
+    {
+        return (new BetterReflection())->constantReflector()->reflect($constantName);
+    }
+
+    /**
+     * Create a reflection of a constant
+     *
+     * @internal
+     *
+     * @param Node\Stmt\Const_|Node\Expr\FuncCall $node Node has to be processed by the PhpParser\NodeVisitor\NameResolver
+     */
+    public static function createFromNode(
+        Reflector $reflector,
+        Node $node,
+        LocatedSource $locatedSource,
+        ?NamespaceNode $namespace = null,
+        ?int $positionInNode = null
+    ) : self {
+        return $node instanceof Node\Stmt\Const_
+            ? self::createFromConstNode($reflector, $node, $locatedSource, $namespace, $positionInNode)
+            : self::createFromFuncCallNode($reflector, $node, $locatedSource, $namespace);
+    }
+
+    private static function createFromConstNode(
+        Reflector $reflector,
+        Node\Stmt\Const_ $node,
+        LocatedSource $locatedSource,
+        ?NamespaceNode $namespace,
+        int $positionInNode
+    ) : self {
+        $constant                     = new self();
+        $constant->reflector          = $reflector;
+        $constant->node               = $node;
+        $constant->locatedSource      = $locatedSource;
+        $constant->declaringNamespace = $namespace;
+        $constant->positionInNode     = $positionInNode;
+
+        return $constant;
+    }
+
+    private static function createFromFuncCallNode(
```

Comment:
S/func/function

Reply:
Is this for `declare()` calls?

Reply:
Yes, it's for `define` calls.

Reply:
Can we rename this to make that obvious?

Reply:
Renamed to `createFromConstKeyword` and `createFromDefineFunctionCall`